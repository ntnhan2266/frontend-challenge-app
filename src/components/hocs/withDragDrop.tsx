import React, { createRef } from "react";
import * as d3 from "d3";
import { ShapeComponentType } from "common/type";

const withDragDrop = (
  WrappedComponent: ShapeComponentType<any>
): typeof React.Component => {
  return class extends React.Component<any | Readonly<any>, any> {
    node = createRef();

    componentDidMount() {
      this.handleDragDropEvent();
    }

    handleDragDropEvent = () => {
      if (this.node.current) {
        const self = this;
        const handleDrag = d3
          .drag()
          .subject(function () {
            const s = d3.select(this);
            return { x: s.attr("cx"), y: s.attr("cy") };
          })
          .on("drag", function (event, _) {
            const s = d3.select(this);
            s.attr("cx", event.x);
            s.attr("cy", event.y);
            self.props.onDrag(WrappedComponent.ShapeType, event.x, event.y);
          });
        handleDrag(d3.select(self.node.current as Element));
      }
    };

    render() {
      return <WrappedComponent {...this.props} domNodeRef={this.node} />;
    }
  };
};

export default withDragDrop;
