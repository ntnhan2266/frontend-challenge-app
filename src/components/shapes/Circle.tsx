import React, { RefObject } from "react";
import "components/shapes/circle.css";
import { SHAPE_TYPE } from "utils/shape";
import withDragDrop from "components/hocs/withDragDrop";
import { ShapeComponentType } from "common/type";

export interface CircleProps {
  cx: number;
  cy: number;
  r: number;
}

interface Props extends CircleProps {
  onDrag: Function;
  isIntersected: boolean;
  domNodeRef: RefObject<any>;
}

export const Circle: React.FunctionComponent<Props> = (props: Props) => {
  return (
    <React.Fragment>
      <circle
        className={props.isIntersected ? "fill-red" : "fill-green"}
        ref={props.domNodeRef}
        cx={props.cx}
        cy={props.cy}
        r={props.r}
      />
    </React.Fragment>
  );
};

const CircleComponent: ShapeComponentType<any> = Object.assign(Circle, {
  ShapeType: SHAPE_TYPE.CIRCLE,
});
export default withDragDrop(CircleComponent);
