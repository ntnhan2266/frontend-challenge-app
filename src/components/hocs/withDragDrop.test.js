import React from "react";
import { mount, shallow } from "enzyme";
import withDragDrop from "components/hocs/withDragDrop";
import Circle from "components/shapes/Circle";

describe("withDragDrop", () => {
  it("should render HOC component correctly", () => {
    const WithDragDropComponent = withDragDrop(Circle);
    const wrapper = shallow(<WithDragDropComponent />);
    expect(wrapper).toMatchSnapshot();
  });
});
