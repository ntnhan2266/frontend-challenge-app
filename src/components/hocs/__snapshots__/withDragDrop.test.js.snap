// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withDragDrop should render HOC component correctly 1`] = `
<_temp
  domNodeRef={
    Object {
      "current": null,
    }
  }
/>
`;
